### DISCUSSION TITLE
How to find what handler ultimately handled a request?

### DISCUSSION THREAD

**Discussion by jackc** [go-chi/chi#828] [Q&A] [answered]
I have a project with a large number of routes. Many of them have URL params. Some of them are grouped into subrouters.

When I have an issue at a particular URL, sometimes it is difficult to determine what handler is responsible.

For example, given the README.md example routes:

```go
 r.Route("/articles", func(r chi.Router) {
    r.With(paginate).Get("/", listArticles)                           // GET /articles
    r.With(paginate).Get("/{month}-{day}-{year}", listArticlesByDate) // GET /articles/01-16-2017

    r.Post("/", createArticle)                                        // POST /articles
    r.Get("/search", searchArticles)                                  // GET /articles/search

    // Regexp url parameters:
    r.Get("/{articleSlug:[a-z-]+}", getArticleBySlug)                // GET /articles/home-is-toronto

    // Subrouters:
    r.Route("/{articleID}", func(r chi.Router) {
      r.Use(ArticleCtx)
      r.Get("/", getArticle)                                          // GET /articles/123
      r.Put("/", updateArticle)                                       // PUT /articles/123
      r.Delete("/", deleteArticle)                                    // DELETE /articles/123
    })
  })
```

How would I easily find that `PUT /articles/123` was routed to the `updateArticle` function? Obviously, I can go path segment by path segment and eventually find it. But when there are longer paths that can take a little while.

I'm somewhat spoiled by the Ruby on Rails logs where it logs the exact method that handled each request.

**Comment by ousloob** [comment]
@jackc
Chi does not log the exact handler function like Rails, but you can log the matched route pattern using `chi.RouteContext()`. This won’t give you the actual function name, but it will show which route definition was matched, including path variable parameters.
```go
chi.RouteContext(r.Context()).RoutePatterns
```

Hope it helps :)

**Comment by ousloob** [comment, marked as answer]
@jackc 
The only solution I could prototype to get the name of the handler is with this:
```go
handlerName := runtime.FuncForPC(reflect.ValueOf(your_handler).Pointer()).Name()
```

However, we can't directly access the original handler function since it's already wrapped by Chi's routing system.

**Reply by jackc** [reply]
Yeah, I was afraid it would end up something like this. The combination of Go's type system and Chi's design makes this awkward at best.

Thanks, I'm going to mark this as the answer as I think this is the best we're going to get.
